File: cpp/app/workload.cpp
#include "workload.hpp"

#include "sha256.hpp"

#include <algorithm>
#include <array>
#include <bit>
#include <cerrno>
#include <cmath>
#include <cstring>
#include <fcntl.h>
#include <limits>
#include <span>
#include <string>
#include <string_view>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <utility>
#include <vector>

namespace anchor::app {
namespace {

constexpr std::size_t kHeaderSize = 128;
constexpr std::uint64_t kAlignment = 64;
constexpr std::array<std::byte, 8> kMagic{
    std::byte{'A'}, std::byte{'N'}, std::byte{'C'}, std::byte{'H'},
    std::byte{'O'}, std::byte{'R'}, std::byte{'W'}, std::byte{0}};

class FileDescriptor {
 public:
  explicit FileDescriptor(const std::filesystem::path& path) {
    descriptor_ = ::open(path.c_str(), O_RDONLY | O_CLOEXEC);
    if (descriptor_ < 0) {
      throw WorkloadFormatError("cannot open workload: " +
                                std::string(std::strerror(errno)));
    }
  }
  FileDescriptor(const FileDescriptor&) = delete;
  FileDescriptor& operator=(const FileDescriptor&) = delete;
  ~FileDescriptor() {
    if (descriptor_ >= 0) ::close(descriptor_);
  }
  [[nodiscard]] int get() const noexcept { return descriptor_; }

 private:
  int descriptor_{-1};
};

std::uint16_t read_le16(const std::byte* source) {
  return static_cast<std::uint16_t>(
      static_cast<std::uint16_t>(std::to_integer<std::uint8_t>(source[0])) |
      (static_cast<std::uint16_t>(std::to_integer<std::uint8_t>(source[1]))
       << 8U));
}

std::uint32_t read_le32(const std::byte* source) {
  std::uint32_t value = 0;
  for (std::size_t i = 0; i < 4; ++i) {
    value |= static_cast<std::uint32_t>(
                 std::to_integer<std::uint8_t>(source[i]))
             << static_cast<unsigned>(8 * i);
  }
  return value;
}

std::uint64_t read_le64(const std::byte* source) {
  std::uint64_t value = 0;
  for (std::size_t i = 0; i < 8; ++i) {
    value |= static_cast<std::uint64_t>(
                 std::to_integer<std::uint8_t>(source[i]))
             << static_cast<unsigned>(8 * i);
  }
  return value;
}

std::uint64_t checked_add64(std::uint64_t left, std::uint64_t right,
                            std::string_view what) {
  if (right > std::numeric_limits<std::uint64_t>::max() - left) {
    throw WorkloadFormatError(std::string(what) + " overflows uint64");
  }
  return left + right;
}

std::uint64_t checked_mul64(std::uint64_t left, std::uint64_t right,
                            std::string_view what) {
  if (left != 0 && right > std::numeric_limits<std::uint64_t>::max() / left) {
    throw WorkloadFormatError(std::string(what) + " overflows uint64");
  }
  return left * right;
}

std::size_t as_size(std::uint64_t value, std::string_view what) {
  if (value > std::numeric_limits<std::size_t>::max()) {
    throw WorkloadFormatError(std::string(what) + " exceeds size_t");
  }
  return static_cast<std::size_t>(value);
}

void pread_exact(int descriptor, std::uint64_t offset,
                 std::span<std::byte> destination, std::string_view what) {
  if (offset > static_cast<std::uint64_t>(
                   std::numeric_limits<off_t>::max())) {
    throw WorkloadFormatError(std::string(what) + " offset exceeds off_t");
  }
  std::size_t completed = 0;
  while (completed < destination.size()) {
    const std::size_t request = std::min<std::size_t>(
        destination.size() - completed,
        static_cast<std::size_t>(std::numeric_limits<ssize_t>::max()));
    const std::uint64_t position =
        checked_add64(offset, static_cast<std::uint64_t>(completed), what);
    if (position > static_cast<std::uint64_t>(
                       std::numeric_limits<off_t>::max())) {
      throw WorkloadFormatError(std::string(what) + " offset exceeds off_t");
    }
    const ssize_t result = ::pread(descriptor, destination.data() + completed,
                                   request, static_cast<off_t>(position));
    if (result < 0) {
      if (errno == EINTR) continue;
      throw WorkloadFormatError(std::string("cannot read ") +
                                std::string(what) + ": " +
                                std::strerror(errno));
    }
    if (result == 0) {
      throw WorkloadFormatError(std::string("truncated ") + std::string(what));
    }
    completed += static_cast<std::size_t>(result);
  }
}

template <class T>
std::vector<T> read_array(int descriptor, std::uint64_t offset,
                          std::uint64_t count, Sha256& digest,
                          std::string_view name) {
  const std::size_t size = as_size(count, name);
  std::vector<T> values(size);
  auto bytes = std::as_writable_bytes(std::span<T>(values));
  pread_exact(descriptor, offset, bytes, name);
  digest.update(std::as_bytes(std::span<const T>(values)));
  return values;
}

void validate_padding(int descriptor, std::uint64_t begin,
                      std::uint64_t end) {
  if (end < begin) throw WorkloadFormatError("overlapping workload arrays");
  std::array<std::byte, 64 * 1024> bytes{};
  std::uint64_t cursor = begin;
  while (cursor < end) {
    const std::uint64_t remaining = end - cursor;
    const std::size_t amount = static_cast<std::size_t>(
        std::min<std::uint64_t>(remaining, bytes.size()));
    auto chunk = std::span<std::byte>(bytes).first(amount);
    pread_exact(descriptor, cursor, chunk, "alignment padding");
    if (std::any_of(chunk.begin(), chunk.end(),
                    [](std::byte value) { return value != std::byte{0}; })) {
      throw WorkloadFormatError("nonzero workload alignment padding");
    }
    cursor = checked_add64(cursor, amount, "padding cursor");
  }
}

template <class Coord>
void validate_positive_finite(std::span<const Coord> lower,
                              std::span<const Coord> upper,
                              std::string_view side) {
  if (lower.size() != upper.size()) {
    throw WorkloadFormatError(std::string(side) + " coordinate size mismatch");
  }
  for (std::size_t i = 0; i < lower.size(); ++i) {
    if constexpr (std::same_as<Coord, double>) {
      if (!std::isfinite(lower[i]) || !std::isfinite(upper[i])) {
        throw NumericDegeneracyError(std::string(side) +
                                     " contains a non-finite endpoint");
      }
    }
    if (!(lower[i] < upper[i])) {
      throw NumericDegeneracyError(std::string(side) +
                                   " contains an empty or inverted box");
    }
  }
}

template <class Coord>
std::variant<DoubleInstance, Int64Instance> make_instance(
    std::size_t dimension, std::vector<std::uint64_t> r_ids,
    std::vector<Coord> r_lower, std::vector<Coord> r_upper,
    std::vector<std::uint64_t> s_ids, std::vector<Coord> s_lower,
    std::vector<Coord> s_upper) {
  validate_positive_finite<Coord>(r_lower, r_upper, "R");
  validate_positive_finite<Coord>(s_lower, s_upper, "S");
  try {
    BoxSet<Coord> r(dimension, std::move(r_ids), std::move(r_lower),
                    std::move(r_upper));
    BoxSet<Coord> s(dimension, std::move(s_ids), std::move(s_lower),
                    std::move(s_upper));
    auto instance = std::make_shared<const BoxJoinInstance<Coord>>(
        std::move(r), std::move(s));
    return std::variant<DoubleInstance, Int64Instance>{std::move(instance)};
  } catch (const std::invalid_argument& error) {
    throw WorkloadFormatError(std::string("invalid workload arrays: ") +
                              error.what());
  }
}

}  // namespace

LoadedWorkload load_workload(const std::filesystem::path& path) {
  if constexpr (std::endian::native != std::endian::little) {
    throw WorkloadFormatError(
        "canonical workload loading requires a little-endian host");
  }

  FileDescriptor file(path);
  struct stat metadata {};
  if (::fstat(file.get(), &metadata) != 0) {
    throw WorkloadFormatError("cannot stat workload: " +
                              std::string(std::strerror(errno)));
  }
  if (!S_ISREG(metadata.st_mode) || metadata.st_size < 0) {
    throw WorkloadFormatError("workload must be a regular file");
  }
  const std::uint64_t actual_size = static_cast<std::uint64_t>(metadata.st_size);
  std::array<std::byte, kHeaderSize> header{};
  pread_exact(file.get(), 0, header, "workload header");
  if (!std::equal(kMagic.begin(), kMagic.end(), header.begin())) {
    throw WorkloadFormatError("invalid workload magic");
  }
  if (read_le16(header.data() + 8) != 1 ||
      std::to_integer<std::uint8_t>(header[10]) != 1 ||
      read_le32(header.data() + 20) != kHeaderSize) {
    throw WorkloadFormatError("unsupported workload format or byte order");
  }
  const auto endpoint_code =
      std::to_integer<std::uint8_t>(header[11]);
  if (endpoint_code != static_cast<std::uint8_t>(EndpointType::Float64) &&
      endpoint_code != static_cast<std::uint8_t>(EndpointType::Int64)) {
    throw WorkloadFormatError("unknown endpoint type code");
  }
  const std::uint64_t dimension = read_le32(header.data() + 12);
  if (dimension == 0 || read_le32(header.data() + 16) != 0) {
    throw WorkloadFormatError("invalid dimension or unsupported header flags");
  }
  const std::uint64_t n_r = read_le64(header.data() + 24);
  const std::uint64_t n_s = read_le64(header.data() + 32);
  std::array<std::uint64_t, 6> offsets{};
  for (std::size_t i = 0; i < offsets.size(); ++i) {
    offsets[i] = read_le64(header.data() + 40 + 8 * i);
  }
  const std::uint64_t declared_size = read_le64(header.data() + 88);
  if (declared_size != actual_size) {
    throw WorkloadFormatError("file size differs from workload header");
  }
  std::array<std::uint8_t, 32> expected_digest{};
  for (std::size_t i = 0; i < expected_digest.size(); ++i) {
    expected_digest[i] = std::to_integer<std::uint8_t>(header[96 + i]);
  }

  const std::uint64_t r_coordinates =
      checked_mul64(n_r, dimension, "R coordinate count");
  const std::uint64_t s_coordinates =
      checked_mul64(n_s, dimension, "S coordinate count");
  const std::array<std::uint64_t, 6> counts{
      n_r, r_coordinates, r_coordinates, n_s, s_coordinates, s_coordinates};
  std::array<std::uint64_t, 6> byte_sizes{};
  std::uint64_t logical_payload_bytes = 0;
  std::uint64_t previous_end = kHeaderSize;
  for (std::size_t i = 0; i < offsets.size(); ++i) {
    byte_sizes[i] = checked_mul64(counts[i], 8, "array byte size");
    if (offsets[i] < previous_end || offsets[i] % kAlignment != 0) {
      throw WorkloadFormatError("array offset is overlapping or unaligned");
    }
    validate_padding(file.get(), previous_end, offsets[i]);
    previous_end = checked_add64(offsets[i], byte_sizes[i], "array end");
    if (previous_end > declared_size) {
      throw WorkloadFormatError("array extends beyond end of workload");
    }
    logical_payload_bytes = checked_add64(
        logical_payload_bytes, byte_sizes[i], "logical payload size");
  }
  if (previous_end != declared_size) {
    throw WorkloadFormatError("unexpected trailing workload bytes");
  }

  Sha256 digest;
  auto r_ids = read_array<std::uint64_t>(file.get(), offsets[0], counts[0],
                                         digest, "R id array");
  const EndpointType endpoint_type = static_cast<EndpointType>(endpoint_code);
  std::variant<DoubleInstance, Int64Instance> instance;
  if (endpoint_type == EndpointType::Float64) {
    auto r_lower = read_array<double>(file.get(), offsets[1], counts[1], digest,
                                      "R lower array");
    auto r_upper = read_array<double>(file.get(), offsets[2], counts[2], digest,
                                      "R upper array");
    auto s_ids = read_array<std::uint64_t>(file.get(), offsets[3], counts[3],
                                           digest, "S id array");
    auto s_lower = read_array<double>(file.get(), offsets[4], counts[4], digest,
                                      "S lower array");
    auto s_upper = read_array<double>(file.get(), offsets[5], counts[5], digest,
                                      "S upper array");
    if (digest.finish() != expected_digest) {
      throw WorkloadFormatError("logical payload SHA-256 mismatch");
    }
    instance = make_instance<double>(
        as_size(dimension, "dimension"), std::move(r_ids), std::move(r_lower),
        std::move(r_upper), std::move(s_ids), std::move(s_lower),
        std::move(s_upper));
  } else {
    auto r_lower = read_array<std::int64_t>(file.get(), offsets[1], counts[1],
                                            digest, "R lower array");
    auto r_upper = read_array<std::int64_t>(file.get(), offsets[2], counts[2],
                                            digest, "R upper array");
    auto s_ids = read_array<std::uint64_t>(file.get(), offsets[3], counts[3],
                                           digest, "S id array");
    auto s_lower = read_array<std::int64_t>(file.get(), offsets[4], counts[4],
                                            digest, "S lower array");
    auto s_upper = read_array<std::int64_t>(file.get(), offsets[5], counts[5],
                                            digest, "S upper array");
    if (digest.finish() != expected_digest) {
      throw WorkloadFormatError("logical payload SHA-256 mismatch");
    }
    instance = make_instance<std::int64_t>(
        as_size(dimension, "dimension"), std::move(r_ids), std::move(r_lower),
        std::move(r_upper), std::move(s_ids), std::move(s_lower),
        std::move(s_upper));
  }
  // Do not keep workload page-cache pages charged to a memory-measurement process.
  (void)::posix_fadvise(file.get(), 0, 0, POSIX_FADV_DONTNEED);

  return LoadedWorkload{path,
                        endpoint_type,
                        as_size(dimension, "dimension"),
                        as_size(n_r, "R object count"),
                        as_size(n_s, "S object count"),
                        declared_size,
                        logical_payload_bytes,
                        expected_digest,
                        std::move(instance)};
}

std::string_view endpoint_type_name(EndpointType type) noexcept {
  switch (type) {
    case EndpointType::Float64:
      return "float64";
    case EndpointType::Int64:
      return "int64";
  }
  return "unknown";
}

}  // namespace anchor::app

File: python/anchor_exp/workload.py
"""Canonical uncompressed workload format.

The file is structure-of-arrays so the benchmark loader can read each component
directly into its final anonymous allocation.  Every array begins at a 64-byte
boundary.  The 128-byte little-endian header is::

    magic[8], version:u16, endian:u8, endpoint:u8,
    dimension:u32, flags:u32, header_size:u32,
    n_R:u64, n_S:u64,
    offsets for R ids/lower/upper and S ids/lower/upper (six u64),
    file_size:u64, logical_payload_sha256[32].

Coordinates are row-major ``(object, dimension)``.  The logical payload digest
covers the six arrays in header order and excludes alignment padding.
"""

from __future__ import annotations

import dataclasses
import hashlib
import json
import os
import pathlib
import struct
from collections.abc import Mapping
from typing import Any

import numpy as np

from .stable_hash import canonical_json_bytes, hash_file


MAGIC = b"ANCHORW\0"
FORMAT_VERSION = 1
ENDIAN_LITTLE = 1
HEADER_SIZE = 128
ALIGNMENT = 64
HEADER = struct.Struct("<8sHBBIII" + "Q" * 9 + "32s")
assert HEADER.size == HEADER_SIZE

ENDPOINT_CODES = {"float64": 1, "int64": 2}
CODE_ENDPOINTS = {value: key for key, value in ENDPOINT_CODES.items()}
ENDPOINT_DTYPES = {"float64": np.dtype("<f8"), "int64": np.dtype("<i8")}
IDS_DTYPE = np.dtype("<u8")


class WorkloadFormatError(ValueError):
    pass


@dataclasses.dataclass(frozen=True)
class Workload:
    path: pathlib.Path
    endpoint_type: str
    dimension: int
    r_ids: np.ndarray
    r_lower: np.ndarray
    r_upper: np.ndarray
    s_ids: np.ndarray
    s_lower: np.ndarray
    s_upper: np.ndarray
    payload_sha256: str
    file_sha256: str | None = None

    @property
    def n_r(self) -> int:
        return int(self.r_ids.size)

    @property
    def n_s(self) -> int:
        return int(self.s_ids.size)


def _align(offset: int) -> int:
    return (offset + ALIGNMENT - 1) // ALIGNMENT * ALIGNMENT


def _as_ids(name: str, values: Any) -> np.ndarray:
    # Object dtype avoids NumPy promoting a mixed small/large Python-int list to
    # float64 before we have checked the full uint64 domain.
    original = np.asarray(values, dtype=object)
    if original.ndim != 1:
        raise ValueError(f"{name} must be one-dimensional")
    integers = original.tolist()
    if any(
        isinstance(value, bool)
        or not isinstance(value, (int, np.integer))
        or int(value) < 0
        or int(value) >= 2**64
        for value in integers
    ):
        raise ValueError(f"{name} values must lie in uint64")
    array = np.ascontiguousarray(integers, dtype=IDS_DTYPE)
    if np.unique(array).size != array.size:
        raise ValueError(f"{name} contains duplicate object ids")
    return array


def _as_coordinates(name: str, values: Any, dtype: np.dtype) -> np.ndarray:
    array = np.ascontiguousarray(values, dtype=dtype)
    if array.ndim != 2:
        raise ValueError(f"{name} must have shape (objects, dimensions)")
    return array


def _validate_arrays(
    *,
    r_ids: Any,
    r_lower: Any,
    r_upper: Any,
    s_ids: Any,
    s_lower: Any,
    s_upper: Any,
    endpoint_type: str,
) -> tuple[np.ndarray, ...]:
    if endpoint_type not in ENDPOINT_DTYPES:
        raise ValueError("endpoint_type must be 'float64' or 'int64'")
    dtype = ENDPOINT_DTYPES[endpoint_type]
    rid = _as_ids("r_ids", r_ids)
    sid = _as_ids("s_ids", s_ids)
    rlo = _as_coordinates("r_lower", r_lower, dtype)
    rhi = _as_coordinates("r_upper", r_upper, dtype)
    slo = _as_coordinates("s_lower", s_lower, dtype)
    shi = _as_coordinates("s_upper", s_upper, dtype)
    if rlo.shape != rhi.shape or slo.shape != shi.shape:
        raise ValueError("lower and upper arrays must have identical shapes")
    if rlo.shape[0] != rid.size or slo.shape[0] != sid.size:
        raise ValueError("id and coordinate object counts differ")
    if rlo.shape[1] != slo.shape[1] or rlo.shape[1] < 1:
        raise ValueError("both sides must use the same positive dimension")
    if endpoint_type == "float64":
        if not all(np.isfinite(a).all() for a in (rlo, rhi, slo, shi)):
            raise ValueError("floating endpoint arrays contain NaN or infinity")
    if not np.all(rlo < rhi) or not np.all(slo < shi):
        raise ValueError("canonical performance workloads cannot contain empty boxes")
    return rid, rlo, rhi, sid, slo, shi


def _array_bytes(array: np.ndarray) -> memoryview:
    return memoryview(array).cast("B")


def _summary(array: np.ndarray) -> list[int | float]:
    if array.shape[0] == 0:
        return []
    values = np.min(array, axis=0)
    return [item.item() for item in values]


def _summary_max(array: np.ndarray) -> list[int | float]:
    if array.shape[0] == 0:
        return []
    values = np.max(array, axis=0)
    return [item.item() for item in values]


def write_workload(
    path: str | pathlib.Path,
    *,
    r_ids: Any,
    r_lower: Any,
    r_upper: Any,
    s_ids: Any,
    s_lower: Any,
    s_upper: Any,
    endpoint_type: str,
    metadata: Mapping[str, Any] | None = None,
    manifest_path: str | pathlib.Path | None = None,
) -> dict[str, Any]:
    """Validate and atomically write a canonical workload and its manifest."""

    arrays = _validate_arrays(
        r_ids=r_ids,
        r_lower=r_lower,
        r_upper=r_upper,
        s_ids=s_ids,
        s_lower=s_lower,
        s_upper=s_upper,
        endpoint_type=endpoint_type,
    )
    rid, rlo, rhi, sid, slo, shi = arrays
    dimension = int(rlo.shape[1])

    offsets: list[int] = []
    cursor = HEADER_SIZE
    for array in arrays:
        cursor = _align(cursor)
        offsets.append(cursor)
        cursor += array.nbytes
    file_size = cursor

    payload_digest = hashlib.sha256()
    for array in arrays:
        payload_digest.update(_array_bytes(array))
    payload_hash = payload_digest.digest()
    header = HEADER.pack(
        MAGIC,
        FORMAT_VERSION,
        ENDIAN_LITTLE,
        ENDPOINT_CODES[endpoint_type],
        dimension,
        0,
        HEADER_SIZE,
        int(rid.size),
        int(sid.size),
        *offsets,
        file_size,
        payload_hash,
    )

    destination = pathlib.Path(path).resolve()
    destination.parent.mkdir(parents=True, exist_ok=True)
    temporary = destination.with_name(f".{destination.name}.tmp-{os.getpid()}")
    try:
        with temporary.open("wb", buffering=0) as stream:
            stream.write(header)
            position = HEADER_SIZE
            for offset, array in zip(offsets, arrays, strict=True):
                if offset > position:
                    stream.write(b"\0" * (offset - position))
                stream.write(_array_bytes(array))
                position = offset + array.nbytes
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(temporary, destination)
    finally:
        if temporary.exists():
            temporary.unlink()

    file_hash = hash_file(destination)
    manifest: dict[str, Any] = {
        "schema_version": "anchor-workload-manifest-v1",
        "format": {
            "magic": MAGIC.rstrip(b"\0").decode("ascii"),
            "version": FORMAT_VERSION,
            "endianness": "little",
            "layout": "soa-row-major-aligned64",
        },
        "workload": {
            "file_name": destination.name,
            "file_size_bytes": file_size,
            "sha256": file_hash,
            "payload_sha256": payload_hash.hex(),
            "endpoint_type": endpoint_type,
            "dimension": dimension,
            "N_total": int(rid.size + sid.size),
            "n_R": int(rid.size),
            "n_S": int(sid.size),
            "R_ids_sha256": hashlib.sha256(_array_bytes(rid)).hexdigest(),
            "S_ids_sha256": hashlib.sha256(_array_bytes(sid)).hexdigest(),
            "coordinate_min": {
                "R": _summary(rlo),
                "S": _summary(slo),
            },
            "coordinate_max": {
                "R": _summary_max(rhi),
                "S": _summary_max(shi),
            },
        },
        "metadata": dict(metadata or {}),
    }
    mpath = (
        pathlib.Path(manifest_path).resolve()
        if manifest_path is not None
        else pathlib.Path(str(destination) + ".manifest.json")
    )
    mpath.parent.mkdir(parents=True, exist_ok=True)
    mtemporary = mpath.with_name(f".{mpath.name}.tmp-{os.getpid()}")
    try:
        with mtemporary.open("wb") as stream:
            stream.write(canonical_json_bytes(manifest) + b"\n")
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(mtemporary, mpath)
    finally:
        if mtemporary.exists():
            mtemporary.unlink()
    return manifest


def _read_header(path: pathlib.Path) -> tuple[Any, ...]:
    with path.open("rb") as stream:
        data = stream.read(HEADER_SIZE)
    if len(data) != HEADER_SIZE:
        raise WorkloadFormatError("truncated workload header")
    fields = HEADER.unpack(data)
    magic, version, endian, endpoint_code, dimension, flags, header_size = fields[:7]
    if magic != MAGIC:
        raise WorkloadFormatError("invalid workload magic")
    if version != FORMAT_VERSION or header_size != HEADER_SIZE:
        raise WorkloadFormatError("unsupported workload format version")
    if endian != ENDIAN_LITTLE:
        raise WorkloadFormatError("only canonical little-endian workloads are supported")
    if endpoint_code not in CODE_ENDPOINTS:
        raise WorkloadFormatError("unknown endpoint type code")
    if dimension < 1 or flags != 0:
        raise WorkloadFormatError("invalid dimension or unsupported header flags")
    return fields


def read_workload(
    path: str | pathlib.Path,
    *,
    verify_payload: bool = True,
    verify_file: bool = False,
) -> Workload:
    """Memory-map a workload; no coordinate copy is made."""

    source = pathlib.Path(path).resolve()
    fields = _read_header(source)
    endpoint_code = fields[3]
    dimension = int(fields[4])
    n_r, n_s = map(int, fields[7:9])
    offsets = tuple(map(int, fields[9:15]))
    declared_size = int(fields[15])
    payload_hash = fields[16]
    actual_size = source.stat().st_size
    if declared_size != actual_size:
        raise WorkloadFormatError("file size differs from header")
    endpoint_type = CODE_ENDPOINTS[endpoint_code]
    dtype = ENDPOINT_DTYPES[endpoint_type]
    shapes = ((n_r,), (n_r, dimension), (n_r, dimension), (n_s,), (n_s, dimension), (n_s, dimension))
    dtypes = (IDS_DTYPE, dtype, dtype, IDS_DTYPE, dtype, dtype)
    arrays: list[np.ndarray] = []
    previous_end = HEADER_SIZE
    for offset, shape, item_dtype in zip(offsets, shapes, dtypes, strict=True):
        if offset % ALIGNMENT or offset < previous_end:
            raise WorkloadFormatError("array offset is unaligned or overlaps its predecessor")
        count = int(np.prod(shape, dtype=np.int64))
        end = offset + count * item_dtype.itemsize
        if end > declared_size:
            raise WorkloadFormatError("array extends beyond end of file")
        arrays.append(np.memmap(source, mode="r", dtype=item_dtype, offset=offset, shape=shape, order="C"))
        previous_end = end
    if previous_end != declared_size:
        raise WorkloadFormatError("unexpected trailing bytes")
    if verify_payload:
        digest = hashlib.sha256()
        for array in arrays:
            digest.update(_array_bytes(array))
        if digest.digest() != payload_hash:
            raise WorkloadFormatError("logical payload SHA-256 mismatch")
    rid, rlo, rhi, sid, slo, shi = arrays
    # Header validation is intentionally repeated here so corrupted files never
    # reach algorithm code even when payload verification is disabled.
    if endpoint_type == "float64" and not all(
        np.isfinite(a).all() for a in (rlo, rhi, slo, shi)
    ):
        raise WorkloadFormatError("non-finite floating endpoint")
    if not np.all(rlo < rhi) or not np.all(slo < shi):
        raise WorkloadFormatError("empty or inverted box in canonical workload")
    return Workload(
        path=source,
        endpoint_type=endpoint_type,
        dimension=dimension,
        r_ids=rid,
        r_lower=rlo,
        r_upper=rhi,
        s_ids=sid,
        s_lower=slo,
        s_upper=shi,
        payload_sha256=payload_hash.hex(),
        file_sha256=hash_file(source) if verify_file else None,
    )


def read_manifest(path: str | pathlib.Path) -> dict[str, Any]:
    manifest_path = pathlib.Path(path)
    if not manifest_path.name.endswith(".json"):
        manifest_path = pathlib.Path(str(manifest_path) + ".manifest.json")
    with manifest_path.open("r", encoding="utf-8") as stream:
        value = json.load(stream)
    if value.get("schema_version") != "anchor-workload-manifest-v1":
        raise WorkloadFormatError("unsupported workload manifest schema")
    return value


def pretouch(workload: Workload) -> int:
    """Read one byte per resident page and return a checksum-like accumulator."""

    page = 4096
    accumulator = 0
    for array in (
        workload.r_ids,
        workload.r_lower,
        workload.r_upper,
        workload.s_ids,
        workload.s_lower,
        workload.s_upper,
    ):
        raw = _array_bytes(array)
        for offset in range(0, len(raw), page):
            accumulator ^= int(raw[offset])
        if raw:
            accumulator ^= int(raw[-1])
    return accumulator
